Python maintenance

jngrad opened the pull request:
Description of changes:
- user guide:
   - cleanup code examples and fix bond breakage documentation (e.g. mode `"bind_centers"` vs. `"bond_centers"`)
   - document ghost particles side-effects on observables
   - where changes were made, wrap text to ~80 chars per line to make git diffs more readable
- python testsuite:
   - reduce code duplication by merging similar tests together and by using `tests_common.py` more often
   - improve code coverage by rewriting assertions
- bug fixes:
   - the incorrectly implemented slab search mode of  `system.analysis.nbhood()` was removed (fixes #4517)
   - `TabulatedNonBonded.is_active()` now returns `False` instead of `None` when the bond is inactive
   - always checks for runtime errors when the cell system changes or when a `ScriptInterface` object is instantiated
- API changes:
   - script interface methods no longer return `True` to signal they didn't raise an exception
- script interface improvements:
   - simplify `System` class and document why the `_Globals` class (now renamed to `_BoxGeometry`) has to exist
   - simplify `ScriptInterfaceHelper` classes using automatically generated member methods
   - simplify several core functions that are included in pxd files

jngrad committed:
doc: Improve user guide and docstring

Fix code samples, spell check and grammar check, cleanup docstrings.

jngrad committed:
core: Simplify code

jngrad committed:
python: Refactor script interface classes

`TabulatedNonBonded.is_active()` now returns False instead of None
when the bond is inactive. Let ScriptInterface classes automatically
generate interface methods. Always checks for runtime errors when
the cell system changes or when a ScriptInterface object is created.
Don't return True from methods that can return void (relic from the
ES_OK / ES_ERROR signaling strategy that has been replaced by Python
exceptions).

jngrad committed:
python: Rewrite System class

jngrad committed:
python: Modernize openGL code

jngrad committed:
python: Remove incorrect slab neighborhood search

The algorithm searched for particles in a cylinder perpendicular
to the plane instead of searching for particles in a slab parallel
to the plane.

jngrad committed:
tests: Merge common tests together

jngrad committed:
tests: Improve code coverage

Re-use existing utility functions in tests_common to reduce code
duplication. Replace unreachable exceptions by unittest assertions.
Fix tolerance of LBGPU test case up to single-precision accuracy.

pkreissl commented on doc/sphinx/advanced_methods.rst:
```diff
@@ -12,36 +12,47 @@ literature before using them.
+  * ``distance`` is the distance between two particles at which the binding is triggered.
+    This cutoff distance, ``1.5`` in the example above, is typically chosen slightly larger
+    than the particle diameter. It is also a good choice for the equilibrium length of the bond.
+  * ``bond_centers`` is the bonded interaction (an instance of :class:`~espressomd.interactions.HarmonicBond`)
```
Does this have to be a `HarmonicBond`? Or is any bonded interaction permissible, in principle?

jngrad replied:
Any bond type is allowed. This sentence was supposed to read "in the example above, an instance of HarmonicBond".

pkreissl commented on doc/sphinx/analysis.rst:
```diff
@@ -90,7 +90,7 @@ Particles in the neighborhood
 :meth:`espressomd.analyze.Analysis.nbhood`
 
-Returns a list of the particle ids of that fall within a given radius of a target position.
+Returns a list of the particle ids that fall within a given radius of a target position.
```
To be precise, grammatically this statement is incorrect. While one likelyt understands the intended meaning, the particle **ids** do not fall within the given radius. Rather should be along these lines: "Returns a list of the ids of particles that fall [...]"

pkreissl commented on doc/sphinx/lb.rst:
```diff
@@ -89,7 +97,9 @@ expert, leave their defaults unchanged. If you do change them, note that they
-set up: ``agrid``, ``tau``, ``visc``, ``dens``. For the other parameters, the following are taken: ``bulk_visc=0``, ``gamma_odd=0``, ``gamma_even=0``, ``ext_force_density=[0,0,0]``.
+set up: ``agrid``, ``tau``, ``visc``, ``dens``. For the other parameters,
+the following are taken: ``bulk_visc=0``, ``gamma_odd=0``, ``gamma_even=0``,
+``ext_force_density=[0,0,0]``.
```
``ext_force_density=[0, 0, 0]``?

pkreissl commented on doc/sphinx/lb.rst:
```diff
@@ -102,8 +112,8 @@ Checkpointing
-(``binary=False``) or binary (``binary=True``) format respectively. The load command
-loads the populations from a checkpoint file written with
+(``binary=False``) or binary (``binary=True``) format respectively.
+The load command loads the populations from a checkpoint file written with
```
The ``lb.load_checkpoint`` command [...]?

pkreissl commented on doc/sphinx/lb.rst:
```diff
@@ -188,7 +200,10 @@ Appending three indices to the ``lb`` object returns an object that represents t
 
-All of these properties can be read and used in further calculations. Only the property ``population`` can be modified. The indices ``x,y,z`` are integers and enumerate the LB nodes in the three directions, starts with 0. To modify ``boundary``, refer to :ref:`Setting up boundary conditions`.
+All of these properties can be read and used in further calculations.
+Only the property ``population`` can be modified. The indices ``x,y,z``
```
``x, y, z`` (with spaces)?

pkreissl commented on src/core/polymer.cpp:
```diff
@@ -67,6 +66,23 @@ template <class RNG> static Utils::Vector3d random_unit_vector(RNG &rng) {
+/** Calculate minimal distance to point.
+ *  Note: Particle handles may be invalidated!
+ *  @param partCfg particle selection
+ *  @param pos  point from which to check for close neighbors
```
Double space after "pos"; although, may just be the default indent before the parameter description... caught my eye, though

jngrad replied:
In doxygen comments, the amount of whitespace between a parameter name and its description is arbitrary. In the past I tried to make it more consistent in the codebase by properly aligning the descriptions, but it's time-consuming. And as soon as a new parameter is introduced whose name is a bit too long, or a `@param[in,out]` is changed to a `@param[in]`, the entire list of `@param` needs to be reformatted, possibly with extra line breaks (which can't be masked with `git diff --color-words` because newlines start with an asterisk). Not really worth the effort, and on some occasions it made the resolution of merge conflicts more difficult.

pkreissl commented on src/python/espressomd/system.pyx:
```diff
@@ -16,55 +16,56 @@
+    """
+    Wrapper class required for technical reasons only.
+
+    When reloading from a checkpoint file, the box length, periodicity and
```
depending on your philosophy, an Oxford-comma after periodicity...

pkreissl commented on src/python/espressomd/system.pyx:
```diff
@@ -76,48 +77,36 @@ cdef class _Globals:
-            if len(_box_l) != 3:
+        def __set__(self, box_l):
+            if len(box_l) != 3:
                 raise ValueError("Box length must be of length 3")
```
mus be **array_like** of length 3?

pkreissl commented on src/python/espressomd/system.pyx:
```diff
@@ -76,48 +77,36 @@ cdef class _Globals:
-                    f"periodicity must be of length 3, got length {len(_periodic)}")
-            mpi_set_periodicity(_periodic[0], _periodic[1], _periodic[2])
-            handle_errors("Error while assigning system periodicity")
+                    f"periodicity must be of length 3, got length {len(periodic)}")
```
maybe also "array_like of lenght 3"

pkreissl commented on src/python/espressomd/utils.pyx:
```diff
@@ -235,6 +234,13 @@ cdef Vector3d make_Vector3d(a):
     return v
 
 
+cdef Vector3i make_Vector3i(a):
```
should there be an assert that ``len(a) == 3``?

jngrad replied:
I don't think it's necessary, the helper functions `make_Vector3i()` and `make_Vector3d()` should only be called on a variable `a` after it passed the sanity check `check_type_or_throw_except(a, 3, float, "...")`.

pkreissl commented on src/python/espressomd/visualization.py:
```diff
@@ -1278,18 +1266,17 @@ def _draw_lb_vel(self):
 
-    # FADE PARTICLE CHARGE COLOR FROM WHITE (q=0) to PLUSCOLOR (q=q_max) RESP
-    # MINUSCOLOR (q=q_min)
+    # FADE PARTICLE CHARGE COLOR FROM WHITE (q=0) to PLUS_COLOR (q=q_max) RESP
```
At some point, we should probably just convert the comments in `visualization.py` to normal capitalization instead of UPPERCASE...

jngrad replied:
done! it took forever

jngrad committed:
python: Apply suggestions from code review

Handle 0-dimensional arrays in type checks. Improve error messages.
Cleanup user guide. Convert uppercase comments to lowercase in the
visualizer and remove comments that are redundant with docstrings.

Co-authored-by: Patrick Kreissl <patrick.kreissl@pa-le.de>

pkreissl reviewed (approved):
Thanks a lot for the nice code improvements and for putting in the extra work of capitalization in the visualizer!

reinaual commented on testsuite/python/linear_momentum_lb.py:
```diff
@@ -92,7 +92,7 @@ class LBGPULinearMomentum(LinearMomentumTest, ut.TestCase):
 
     lb_class = espressomd.lb.LBFluidGPU
-    atol = 1e-6
+    atol = 1e-5
```
what happened here that the tolerance needs to be increased?

jngrad replied:
This test is failing at random in CI, e.g. [logfile 317102](https://gitlab.icp.uni-stuttgart.de/espressomd/espresso/-/jobs/317102) in the long-range actors refactoring PR. The numerical error is random with a probability distribution that looks Chi-squared. The test fixes all random seeds and the core is supposed to sum the densities in a deterministic order, so I'm not sure what the root cause is. I ran that test in a loop for several thousand times and the largest deviation was 4e-6.
